README
Digest-SipHash

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Digest::SipHash

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Digest-SipHash

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Digest-SipHash

    CPAN Ratings
        http://cpanratings.perl.org/d/Digest-SipHash

    Search CPAN
        http://search.cpan.org/dist/Digest-SipHash/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Dan Kogai

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
#
#  Revision history for Digest::SipHash
#
#
$Revision: 0.19 $ $Date: 2013/03/05 09:08:12 $
! csiphash.c
  if __BYTE_ORDER__ is not predefined (like FreeBSD), use arch

0.18 2013/03/05 06:52:11
! t/0?-*.t
  More compatible with older perls
! SipHash.xs lib/Digest/SipHash.pm
  No more unpack()ing -- directly passes UV
! csiphash.c
  Stricter little-endian checking.
  The previous version falsely reported PowerBook G4 is little-endian :-(

0.17 2013/03/02 03:15:44
! csiphash.c
  Reverted _le64toh to that of 0.12 - more portable than <endian.h> detection.

0.16 2013/02/28 03:18:03
! lib/Digest/SipHash.pm lib/Digest/SipHash/PP.pm
  Addressed: use constant { ... } causing 'Odd number of elements
  in anonymous hash' errors on some platforms.

0.15 2013/02/27 04:05:09
! csiphash.c
  Fixed *BSD* lines that were corrupt

0.14 2013/02/26 04:24:09
! lib/Digest/SipHash.pm lib/Digest/SipHash/PP.pm
  64bit support detection no longer relies on Config.pm.
  (the logic used in Data::MessagePack::PP)
! csiphash.c
  Fixed macros that was causing OpenBSD builds to fail.

0.13 2013/02/24 05:51:56
! SipHash.xs
  removed unnneeded le64toh()
! csiphash.c
  New one pulled from https://github.com/majek/csiphash

0.12 2013/02/19 01:38:07
! csiphash.c
  s/le646oh/le64toh/ ; /* shi6! */
  (un)?fortunately this bug is very unlikely to surface.
  Your system must be big-endian AND lacks __builtin_bswap64().

0.11 2013/02/18 23:17:48
! csiphash.c
  le64toh() is now more portable.
  Thank Thomas Eckardt for reporting 0.10 was incompatble w/ Windows
  because of that

0.10 2013/02/18 10:43:55
! lib/Digest/SipHash/PP.pm
  instead of $mbi->numify(), 0 + $mbi is used to retrive UV.
  The former does not work reliabily on older perl.
  Thank you, travis!

0.09 2013/02/18 10:20:52
+ lib/Digest/SipHash/PP.pm t/02-PP.t
  Pure-Perl Implementation!
! lib/Digest/SipHash.pm
  POD fixes.

0.08 2013/02/17 15:07:36
! lib/Digest/SipHash.pm t/01-SipHash.t
  + Added siphash32 to @EXPORT_OK
  + Added :all to @EXPORT_TAG

0.07 2013/02/17 14:48:20
! SipHash.xs lib/Digest/SipHash.pm
  + added siphash32(), just an alias of siphash().
  ! s/key/seed/g
    # so less confusing from perl's point of view

0.06 2013/02/17 14:03:18
! lib/Digest/SipHash.pm t/01-SipHash.t
  Fixed: siphash() in scalar context

0.05 2013/02/17 13:34:32
! lib/Digest/SipHash.pm
  POD fixes.  Functionally identical to the previous version

0.04 2013/02/17 13:23:00
! SipHash.xs lib/Digest/SipHash.pm t/01-SipHash.t
  ! siphash() now always returns lower 32-bit first so
  Hash::Util::hash_value($str)
    == Digest::SipHash::siphash($str, Hash::Util::hash_seed());
  always holds true.
  + siphash64() added for 64-bit platforms

0.03 2013/02/17 10:30:45
! SipHash.xs lib/Digest/SipHash.pm t/01-SipHash.t
  Supports 64-bit int if perl supports that.

0.02 2013/02/17 08:30:18
! t/01-SipHash.t
  Checks the bug below.
! lib/Digest/SipHash.pm
  # forgot the following. oops.
  use base 'Exporter'; 

0.01 2013/02/17 07:53:54
+ *
  First Release


lib/Digest/SipHash/PP.pm
package Digest::SipHash::PP;
use strict;
use warnings;
use Math::BigInt;

our $VERSION = sprintf "%d.%02d", q$Revision: 0.12 $ =~ /(\d+)/g;
use base 'Exporter';
our @EXPORT_OK = qw/siphash siphash32/;

use constant USE64BITINT => eval { pack 'Q', 1 };
use constant MASK64 => Math::BigInt->new('0xffff_ffff_ffff_ffff');

push @EXPORT_OK, 'siphash64' if USE64BITINT;
our %EXPORT_TAGS = ( all => [@EXPORT_OK] );
our $DEFAULT_SEED = pack 'C16', map { int( rand(256) ) } ( 0 .. 0xF );

sub _new { Math::BigInt->new(shift) }
sub _u64 { _new($_[1])->blsft(32)->bior($_[0]) }
sub _add { $_[0]->badd($_[1])->band(MASK64) }
sub _xor { $_[0]->bxor($_[1]) }
sub _rot {
    my $lo = $_[0]->copy->brsft(64 - $_[1]);
    $_[0]->blsft($_[1])->bior($lo)->band(MASK64);
}

sub _compress {
    _add( $_[0], $_[1] );
    _add( $_[2], $_[3] );
    _rot( $_[1], 13 );
    _rot( $_[3], 16 );
    _xor( $_[1], $_[0] );
    _xor( $_[3], $_[2] );
    _rot( $_[0], 32 );
    _add( $_[2], $_[1] );
    _add( $_[0], $_[3] );
    _rot( $_[1], 17 );
    _rot( $_[3], 21 );
    _xor( $_[1], $_[2] );
    _xor( $_[3], $_[0] );
    _rot( $_[2], 32 );
}

sub _digest {
    use bytes;
    my $str  = shift || '';
    my $seed = shift || "\0" x 16;
    my @k = unpack 'V4', $seed;
    my $k0 = _u64( @k[ 0, 1 ] );
    my $k1 = _u64( @k[ 2, 3 ] );
    my $v0 = _new('0x736f6d6570736575');
    my $v1 = _new('0x646f72616e646f6d');
    my $v2 = _new('0x6c7967656e657261');
    my $v3 = _new('0x7465646279746573');
    _xor( $v0, $k0 );
    _xor( $v1, $k1 );
    _xor( $v2, $k0 );
    _xor( $v3, $k1 );
    my $slen = length($str);
    $str .= "\0" x ( 7 - $slen % 8 ) . chr( $slen % 256 );
    my @u32 = unpack 'V*', $str;

    while ( my ( $lo, $hi ) = splice @u32, 0, 2 ) {
        my $u64 = _u64( $lo, $hi );
        _xor( $v3, $u64 );
        _compress( $v0, $v1, $v2, $v3 ) for 0 .. 1;
        _xor( $v0, $u64 );
    }
    _xor( $v2, 0xff );
    _compress( $v0, $v1, $v2, $v3 ) for 0 .. 3;
    _xor( _xor( $v0, $v1 ), _xor( $v2, $v3 ) );
}

sub siphash {
    my $u64 = _digest(@_);
    my $lo  = 0 + $u64->copy->band(0xffff_ffff);
    return $lo unless wantarray;
    my $hi  = 0 + $u64->brsft(32);
    return ( $lo, $hi );
}

*siphash32 = \&siphash;

if (USE64BITINT) {
    *siphash64 = sub {
        use integer;
        0+_digest(@_);
    };
}

1;

__END__

=head1 NAME

Digest::SipHash::PP - Pure-Perl implementation of the SipHash algorithm

=head1 VERSION

$Id: PP.pm,v 0.12 2013/02/28 03:18:03 dankogai Exp dankogai $

=head1 SYNOPSIS

  use Digest::SipHash::PP qw/siphash/;
  my $seed = pack 'C16', 0 .. 0xF;    # 16 chars long
  my $str = "hello world!";
  my ( $lo, $hi ) = siphash( $str, $seed );
  #  $lo = 0x10cf32e0, $hi == 0x7da9cd17
  my $u32 = siphash( $str, $seed )
  #  $u32 = 0x10cf32e0

  use Config;
  if ( $Config{use64bitint} ) {
    use Digest::SipHash qw/siphash64/;
    my $uint64 = siphash64( $str, $seed );    # scalar context;
    # $uint64 == 0x7da9cd1710cf32e0
  }

=head1 DESCRIPTION

This module is identical to L<Digest::SipHash> except implementation.
This module is not meant to be practical; written just for the sake of
curiosity.

=head2 IMPLEMENTATION

For the sake of 32-bit compatibility, this module uses L<Math::BigInt>
for 64-bit operations on which SipHash heavily relies.

=head1 EXPORT

C<siphash()>, C<siphash32()> and C<siphash64()> on demand.

C<:all> to all of above

=head1 SUBROUTINES/METHODS

=head2 siphash

=head2 siphash32

=head2 siphash64

See L<Digest::SipHash> for details.

=head1 AUTHOR

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Dan Kogai.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

=cut


lib/Digest/SipHash.pm
package Digest::SipHash;

use 5.008001;
use strict;
use warnings;

our $VERSION = sprintf "%d.%02d", q$Revision: 0.19 $ =~ /(\d+)/g;
require XSLoader;
XSLoader::load( 'Digest::SipHash', $VERSION );

use base 'Exporter';
our @EXPORT_OK = qw/siphash siphash32/;

use constant BIG_ENDIAN => pack( "L", 1 ) eq pack( "N", 1 );
use constant USE64BITINT => eval { pack 'Q', 1 };

push @EXPORT_OK, 'siphash64' if USE64BITINT;
our %EXPORT_TAGS = ( all => [@EXPORT_OK] );
our $DEFAULT_SEED = pack 'C16', map { int( rand(256) ) } ( 0 .. 0xF );

sub siphash {
    my $str = shift;
    my $seed = shift || $DEFAULT_SEED;
    use bytes;
    $seed .= substr( $DEFAULT_SEED, length($seed) ) if length($seed) < 16;
    my $lohi = _xs_siphash_av( $str, $seed );
    return wantarray ? @$lohi : $lohi->[0];
}

*siphash32 = \&siphash;

if (USE64BITINT) {
    *siphash64 = sub {
        my $str = shift;
        my $seed = shift || $DEFAULT_SEED;
        use bytes;
        $seed .= substr( $DEFAULT_SEED, length($seed) ) if length($seed) < 16;
        return _xs_siphash64( $str, $seed );
    };
}

1;

=head1 NAME

Digest::SipHash - Perl XS interface to the SipHash algorithm

=head1 VERSION

$Id: SipHash.pm,v 0.19 2013/03/05 09:07:53 dankogai Exp $

=head1 SYNOPSIS

  use Digest::SipHash qw/siphash/;
  my $seed = pack 'C16', 0 .. 0xF;    # 16 chars long
  my $str = "hello world!";
  my ( $lo, $hi ) = siphash( $str, $seed );
  #  $lo = 0x10cf32e0, $hi == 0x7da9cd17
  my $u32 = siphash( $str, $seed )
  #  $u32 = 0x10cf32e0

  use Config;
  if ( $Config{use64bitint} ) {
    use Digest::SipHash qw/siphash64/;
    my $uint64 = siphash64( $str, $seed );    # scalar context;
    # $uint64 == 0x7da9cd1710cf32e0
  }

=head1 DESCRIPTION

SipHash is the default perl hash function for 64 bit builds now.

L<http://perl5.git.perl.org/perl.git/commit/3db6cbfca39da94d152d3e860e2aa79b9c6bb161>

L<https://131002.net/siphash/>

This module does only one thing - culculates the SipHash value of the
given string.

=head1 EXPORT

C<siphash()>, C<siphash32()> and C<siphash64()> on demand.

C<:all> to all of above

=head1 SUBROUTINES/METHODS

=head2 siphash

  my ($hi, $lo) = siphash($str [, $seed]);
  my $uint32    = siphash($str [, $seed]);

Calculates the SipHash value of C<$src> with $<$seed>.

If C<$seed> is omitted, it defaults to C<$Digest:::SipHash::DEFAULT_SEED>,
which is set randomly upon initialization of this module.

If C<$seed> is set but less than 16 bytes long, it is padded with C<$DEFAULT_SEED>.

To be compatible with 32-bit perl, It returns a pair of 32-bit
integers instead of a 64-bit integer.  Since C<Hash::Util::hash_value()>
always returns the lower 32-bit first so that:

  use Hash::Util qw/hash_seed hash_value/;
  use Digest::SipHash qw/siphash/;
  hash_value($str) == siphash($str, hash_seed()); # scalar context

always holds true when PERL_HASH_FUN_SIPHASH is in effect.

=head2 siphash32

just an alias of C<siphash>.

=head2 siphash64

  my $uint64 = siphash64($str [, $seed]);

Calculates the SipHash value of C<$src> with C<$seed> in 64-bit.
Available on 64-bit platforms only.

=head1 AUTHOR

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-digest-siphash at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Digest-SipHash>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Digest::SipHash

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Digest-SipHash>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Digest-SipHash>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Digest-SipHash>

=item * Search CPAN

L<http://search.cpan.org/dist/Digest-SipHash/>

=back

=head1 SEE ALSO

L<Hash::Util>, L<https://131002.net/siphash/>

=head1 ACKNOWLEDGEMENTS

B<SipHash: a fast short-input PRF>

by Jean-Philippe Aumasson & Daniel J. Bernstein

L<https://131002.net/siphash/>

=head1 LICENSE AND COPYRIGHT

=head2 csiphash.c

Copyright (c) 2013  Marek Majkowski

MIT License L<http://opensource.org/licenses/MIT>

L<https://github.com/majek/csiphash>

=head2 The rest of this module

Copyright 2013 Dan Kogai.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Digest::SipHash
